python/ray/dag/dag_node_operation.py

```diff
@@ -244,32 +197,24 @@ def _add_edge(
         to_node: The node to which the edge points.
         label: The label of the edge. This will be used to annotate the edge
             in the visualization of the execution schedule.
+        control_dependency: Whether the edge represents a control dependency.
     """
-    from_node.out_edges[(to_node.task_idx, to_node.operation.type)] = (
+    from_node.out_edges[to_node.task_idx] = (
         label,
         control_dependency,
     )
-    to_node.in_edges[(from_node.task_idx, from_node.operation.type)] = (
+    to_node.in_edges[from_node.task_idx] = (
         label,
         control_dependency,
     )
 
 
 def _push_candidate_node_if_ready(
     actor_to_candidates: Dict["ray._raylet.ActorID", List[_DAGOperationGraphNode]],
-    graph: Dict[int, Dict[_DAGNodeOperationType, _DAGOperationGraphNode]],
+    graph: Dict[int, _DAGOperationGraphNode],
     node: _DAGOperationGraphNode,
 ) -> None:
-    # Collective operations are ready when all the collective nodes have zero
-    # in-degrees. Only one node per collective will be added as ready.
-    if node.is_nccl_collective:
-        for collective_node_metadata in node.collective_idxs:
-            task_idx, op_type = collective_node_metadata
-            collective_node = graph[task_idx][op_type]
-            collective_node.ready_collective_idxs.add(
-                (node.task_idx, node.operation.type)
-            )
-    if node.is_ready:
+    if node.is_ready(graph):
         heapq.heappush(
```

Comment:
As discussed offline with Weixin, only push once per group, add a flag to indicate whether a `sync_group` has been pushed.

Reply:
Done.